Master technical onboarding faavc

Faruk372742 opened the pull request.

Faruk372742 committed:
[ADD] estate: initialize the estate app - Chapter 2

Creating the manifest file.

Faruk372742 committed:
[ADD] estate: adding estate property table - Chapter 3

Faruk372742 committed:
[ADD] estate: adding security for estate_property table - Chapter 4

New ir.model.access file to adding all kind of accesses to the base.group_user and connecting csv file to manifest

[LINT] estate: fixing linter issues

[LINT] estate: adding new lines to the end of the few file

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1071.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1071)

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,20 @@
+    garden = fields.Boolean()
+    garden_area = fields.Integer()
+    garden_orientation = fields.Selection(string="Orientation", selection=[("North", "North"), ("South", "South"), ("East", "East"), ("West", "West")])
+    
```
do not forget the new line at the end of file

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,20 @@
+    garage = fields.Boolean()
+    garden = fields.Boolean()
+    garden_area = fields.Integer()
+    garden_orientation = fields.Selection(string="Orientation", selection=[("North", "North"), ("South", "South"), ("East", "East"), ("West", "West")])
```
I would prefer to right the selection options that way to make it more clearer Also the tuple first value represent the technical name so it should be in snake_case 🐍 
```suggestion
    garden_orientation = fields.Selection([
        ("north", "North"),
        ("south", "South"),
        ("east", "East"),
        ("west", "West")
    ], string="Orientation")
```

Faruk372742 committed:
[ADD] estate: adding menus and changing fields - Chapter 5

[LINT] estate: linting all python files with ruff

Faruk372742 committed:
[IMP] estate: adding list,form, search views (filters + group by) - Chapter 6

add: state field to the estate_property model
security: readonly and all-rights security groups are created
views: list, form, search views are created, in search view state filter and postcode groupby features are added
manifest is updated with the security file

Faruk372742 committed:
[IMP] estate: adding relations between models - Chapter 7

new models: estate_property_offer, estate_property_tag
estate_property: adding new fields (type, salesperson, buyer, tags, offers)
estate_property_view: adding offers to form view, tags to both list and form views
estate_property_offer_view: adding list and form views
menu: settings menu for types and tags

Faruk372742 committed:
[IMP] estate: add computed fields and onchanges - Chapter 8

computed fields: deadline of the offer, best price from many offers
onchange: garden affects garden_area and garden_orientation

Faruk372742 committed:
[IMP] estate: adding actions to estate property and estate_property_offers - Chapter 9

estate_property: adding cancel and sold buttons and actions for those buttons
estate_property_offer: adding accept and refuse buttons and actions for those buttons
learned: actions, exception example, icons in buttons

Faruk372742 committed:
[IMP] estate: add SQL and Python constraints - Chapter 10

SQL constraints: expected_price, selling_price, offer price, property tag and type
Python constraints: selling price >= 0.9 of the expected price

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,93 @@
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date(
+        copy=False, default=datetime.now() + relativedelta(months=3)
```
for the default values it is get evaluated on server boot-up So without the `lambda` once you run the server it will take the date and after 1 week if you created new property it will take the 3 month from the server boot-up not 3 month from real today.
```suggestion
        copy=False, default=lambda self: datetime.now() + relativedelta(months=3)
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,93 @@
+    buyer_id = fields.Many2one("res.partner", string="Buyer", copy=False)
+    tag_ids = fields.Many2many("estate_property_tag")
+    offer_ids = fields.One2many("estate_property_offer", "property_id", string="Offers")
+    total_area = fields.Integer("Total Area (sqm)", compute="_compute_area")
```
for compute method we tend to use the full field name if it is the only field computed in the method
```suggestion
    total_area = fields.Integer("Total Area (sqm)", compute="_compute_total_area")
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,93 @@
+                0
+                if len(record.offer_ids) == 0
+                else max(offer.price for offer in record.offer_ids)
+            )
```
the `.mapped('field')` return a list with the all the field values for the recordset and the `or [0]` will replace the list of prices with list contain 0 if the list is empty list
```suggestion
            record.best_price = max(record.offer_ids.mapped('price') or [0])
```

amah-odoo reviewed (commented):
Huge work so far 🎉 

just small comment for now

Faruk372742 committed:
[REF] estate: resolving onboarder's comments

ref: maximizing method and name of the computation method

Faruk372742 committed:
[IMP] estate: adding sprinkles to the views - Chapter 11

inline property list view to property type
statusbar widget for the status of the property
manual and model ordering
adding attributes and options (color, invisible, readonly), color decorations
default availability filter, filter_domain for living_area
stat_button for property types that shows offer for a specific property type

Faruk372742 committed:
[IMP] estate: res.user inheritance for properties - Chapter 12

CRUD inheritance (ondelete for offer, create for estate property)
inherit res.user and add property_ids to it
view inheritance: from res.user and add properties of the user to the new page in the form view

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,136 @@
+        default="new",
+        required=True,
+        copy=False,
+        store=True,
```
we do not need this as this is not computed anymore 👀

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,136 @@
+        "CHECK(selling_price > 0)", "The selling price must be positive"
+    )
+
+    _order = "id desc"
```
let's keep all the meta data at first before any field declaration

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,136 @@
+    #     for record in self:
+    #         if len(record.offer_ids) == 1 and record.state == "new":
+    #             # record.state = "offer_received" OPEN here do not forget!!!
+    #             record.state = record.state
```
let's drop this also 🤔

amah-odoo reviewed (commented):
small comments to keep structure 👀

Faruk372742 committed:
[REF] estate: removing comments and endlining xmls

Faruk372742 committed:
[IMP] estate, estate_account: new link module and invoice creation - Chapter 13

estate_account module is created that depends on account and estate
override: selling property action to create invoice
learned: invoice creation basics

Faruk372742 committed:
[IMP] estate: kanban view for estate property - Chapter 14

Faruk372742 committed:
[REF] estate: delete unused parts

amah-odoo reviewed (commented):
Finilizing you framework journey please adapt the code to follow those [Coding Guidelines](https://www.odoo.com/documentation/18.0/contributing/development/coding_guidelines.html#symbols-and-conventions) 🫡 

Thanks for your work 🎉

Faruk372742 committed:
[IMP] awesome_owl: finish chapter 1, learn owl
